src/main/kotlin/io/libp2p/pubsub/gossip/GossipRouter.kt

```diff
@@ -98,104 +235,213 @@ open class GossipRouter : AbstractRouter() {
     }
 
-    override fun subscribe(topic: String) {
+    override fun subscribe(topic: Topic) {
         super.subscribe(topic)
-        val fanoutPeers = fanout[topic] ?: mutableSetOf()
+        val fanoutPeers = (fanout[topic] ?: mutableSetOf())
+            .filter { score.score(it) >= 0 && !isDirect(it) }
         val meshPeers = mesh.getOrPut(topic) { mutableSetOf() }
-        val otherPeers = getTopicPeers(topic) - meshPeers - fanoutPeers
-        if (meshPeers.size < D) {
-            val addFromFanout = fanoutPeers.shuffled(random).take(D - meshPeers.size)
-            val addFromOthers = otherPeers.shuffled(random).take(D - meshPeers.size - addFromFanout.size)
+        val otherPeers = (getTopicPeers(topic) - meshPeers - fanoutPeers)
+            .filter { score.score(it) >= 0 && !isDirect(it) }
+
+        if (meshPeers.size < coreParams.D) {
+            val addFromFanout = fanoutPeers.shuffled(random)
+                .take(coreParams.D - meshPeers.size)
+            val addFromOthers = otherPeers.shuffled(random)
+                .take(coreParams.D - meshPeers.size - addFromFanout.size)
 
-            meshPeers += (addFromFanout + addFromOthers)
             (addFromFanout + addFromOthers).forEach {
                 graft(it, topic)
             }
+            fanout -= topic
+            lastPublished -= topic
         }
     }
 
-    override fun unsubscribe(topic: String) {
+    override fun unsubscribe(topic: Topic) {
         super.unsubscribe(topic)
-        mesh.remove(topic)?.forEach { prune(it, topic) }
+        mesh[topic]?.forEach { prune(it, topic) }
     }
 
-    private fun heartBeat(time: Long) {
+    private fun heartBeat() {
+        heartbeatsCount++
+        iAsked.clear()
+        peerIHave.clear()
+
+        val staleIWantTime = this.curTime() - params.iWantFollowupTime.toMillis()
+        iWantRequests.entries.removeIf { (key, time) ->
+            (time < staleIWantTime)
+                .whenTrue { notifyRouterMisbehavior(key.first, 1) }
+        }
+
         try {
             mesh.entries.forEach { (topic, peers) ->
-                if (peers.size < DLow) {
-                    (getTopicPeers(topic) - peers).shuffled(random).take(D - peers.size).forEach { newPeer ->
-                        peers += newPeer
-                        graft(newPeer, topic)
-                    }
-                } else if (peers.size > DHigh) {
-                    peers.shuffled(random).take(peers.size - D).forEach { dropPeer ->
-                        peers -= dropPeer
-                        prune(dropPeer, topic)
+
+                // drop underscored peers from mesh
+                peers.filter { score.score(it) < 0 }
+                    .forEach { prune(it, topic) }
+
+                if (peers.size < coreParams.DLow) {
+                    // need more mesh peers
+                    (getTopicPeers(topic) - peers)
+                        .filter { score.score(it) >= 0 && !isDirect(it) && !isBackOff(it, topic) }
```

Comment:
I know the spec does not say anything about this, but I'm curious if we should, on top of ignoring peers with a negative score, favor peers with a better score.

Reply:
Do you mean sort the peers by score before taking desired number of peers? 
There is no such requirement and the reference Go implementation doesn't do it: https://github.com/libp2p/go-libp2p-pubsub/blob/master/gossipsub.go#L1283

I believe there could be some sense behind this: we don't want total score biasing, there should be some randomness. But I hope @vyzo might want to comment on this

Reply:
Yes indeed, we want random selection!